src/comgr.cpp

```diff
@@ -0,0 +1,503 @@
+        CASE_TO_STRING(AMD_COMGR_ACTION_ADD_DEVICE_LIBRARIES);
+        CASE_TO_STRING(AMD_COMGR_ACTION_LINK_BC_TO_BC);
+        CASE_TO_STRING(AMD_COMGR_ACTION_OPTIMIZE_BC_TO_BC);
+        CASE_TO_STRING(AMD_COMGR_ACTION_CODEGEN_BC_TO_RELOCATABLE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_CODEGEN_BC_TO_ASSEMBLY);
+        CASE_TO_STRING(AMD_COMGR_ACTION_LINK_RELOCATABLE_TO_RELOCATABLE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_LINK_RELOCATABLE_TO_EXECUTABLE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_ASSEMBLE_SOURCE_TO_RELOCATABLE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_DISASSEMBLE_RELOCATABLE_TO_SOURCE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_DISASSEMBLE_EXECUTABLE_TO_SOURCE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_DISASSEMBLE_BYTES_TO_SOURCE);
+        CASE_TO_STRING(AMD_COMGR_ACTION_COMPILE_SOURCE_TO_FATBIN);
+    }
+    return "<Unknown action kind>";
+}
+
+static bool PrintVersion()
+{
+    std::size_t major = 0;
+    std::size_t minor = 0;
+    (void)amd_comgr_get_version(&major, &minor);
+    MIOPEN_LOG_I("comgr v." << major << '.' << minor);
+    return true;
+}
+
+static std::string GetStatusText(const amd_comgr_status_t status)
+{
+    const char* reason = nullptr;
+    if(AMD_COMGR_STATUS_SUCCESS != amd_comgr_status_string(status, &reason))
+        reason = "<Unknown>";
+    return std::string(reason) + " (" + std::to_string(static_cast<int>(status)) + ')';
+}
+
+static void LogOptions(const char* options[], size_t count)
+{
+#if DEBUG_DETAILED_LOG
+    if(miopen::IsLogging(miopen::LoggingLevel::Info))
+    {
+        std::ostringstream oss;
+        for(std::size_t i = 0; i < count; ++i)
+            oss << options[i] << '\t';
+        MIOPEN_LOG_I(oss.str());
+    }
+#else
+    std::ignore = options;
+    std::ignore = count;
+#endif
+}
+
+class Dataset;
+static std::string GetLog(const Dataset& dataset, bool catch_comgr_exceptions = false);
+
+struct ComgrError : std::exception
+{
+    amd_comgr_status_t status;
+    std::string text;
+
+    ComgrError(const amd_comgr_status_t s) : status(s) {}
+    ComgrError(const amd_comgr_status_t s, const std::string& t) : status(s), text(t) {}
+    const char* what() const noexcept override { return text.c_str(); }
+};
+
+struct ComgrOwner
+{
+    ComgrOwner(const ComgrOwner&) = delete;
+
+    protected:
+    ComgrOwner() {}
+    ComgrOwner(ComgrOwner&&) = default;
+    [[noreturn]] void Throw(const amd_comgr_status_t s) const { throw ComgrError{s}; }
+    [[noreturn]] void Throw(const amd_comgr_status_t s, const std::string& text) const
+    {
+        throw ComgrError{s, text};
+    }
+};
+
+class Data : ComgrOwner
+{
+    amd_comgr_data_t handle = {0};
+    friend class Dataset; // for GetData
+    Data(amd_comgr_data_t h) : handle(h) {}
+
+    public:
+    Data(amd_comgr_data_kind_t kind) { ECI_THROW(amd_comgr_create_data(kind, &handle), kind); }
+    Data(Data&&) = default;
+    ~Data() { EC(amd_comgr_release_data(handle)); }
```

Comment:
Use `MIOPEN_MANAGE_PTR` for this instead of the raw pointer. This will make the class properly movable, and be upgraded to a shared pointer if we need the class to be copyable.

Reply:
I've considered this option when wrote this code and considered this would be an over-design. Let's do that when/if need arise.

Reply:
Its not over-design. This is what we use in miopen for other API objects from opencl, hip, sqlite, etc. And it would simplify what you are writing.

Your class is incomplete(missing assignment operators, noexcept, handling null cleanup, etc). Instead of trying to make it complete just reuse what we already have which is not only complete but is well tested. There is no reason to reinvent the wheel here.

Reply:
How do I pass and check the return code from deleter?

Reply:
`noexcept` optimization is not important in the context of compilation, which is very long operation. `operator=()` is deleted by default in C++11. If I understand correctly, null cleanup is not applicable here. 

"Reinvent the whell" is a value judgment. One can say that using smart pointers + macros + template classes to handle non-pointer objects is cluttering and misleading, but this would be a value judgment too ;) I am just trying to use the simplest things, specifically -- the guard idiom and very basic notions that constitute the core of C++.

Of course I am not totally against of this `MIOPEN_MANAGE_PTR` thing, I am attempting to be as simple as possible _and_ find some grounding for every complication.

Reply:
> One can say that using smart pointers + macros + template classes to handle non-pointer objects is cluttering and misleading,

I think I see. The issue is that `amd_comgr_data_t` is not a pointer but just a handle. `MIOPEN_MANAGE_PTR` doesn't support using a handle like that.